src/ray/core_worker/task_event_buffer.cc

```diff
@@ -143,72 +136,79 @@ void TaskEventBufferImpl::FlushEvents(bool forced) {
       return;
     }
 
-    if (buffer_.size() == 0) {
+    // No data to send.
+    if (buffer_.empty()) {
       return;
     }
 
-    task_events.reserve(
-        RayConfig::instance().task_events_max_num_task_events_in_buffer());
-    buffer_.swap(task_events);
+    if (static_cast<int64_t>(buffer_.size()) >
+        RayConfig::instance().task_events_send_batch_size()) {
+      // We will take a batch from the buffer to send.
+      size_t batch_size = RayConfig::instance().task_events_send_batch_size();
+
+      auto move_start = std::prev(buffer_.end(), batch_size);
+      to_send.insert(to_send.end(),
+                     std::make_move_iterator(move_start),
+                     std::make_move_iterator(buffer_.end()));
+      buffer_.erase(move_start, buffer_.end());
+      // Enough space in the buffer now, reset the next_idx_to_overwrite.
+    } else {
+      // Just send all.
+      to_send.swap(buffer_);
+    }
     next_idx_to_overwrite_ = 0;
 
+    // Send and reset the counters
     num_profile_task_events_dropped = num_profile_task_events_dropped_;
     num_profile_task_events_dropped_ = 0;
 
     num_status_task_events_dropped = num_status_task_events_dropped_;
     num_status_task_events_dropped_ = 0;
   }
 
-  // Merge multiple events from a single task attempt run into one task event.
-  absl::flat_hash_map<std::pair<std::string, int>, rpc::TaskEvents> task_events_map;
+  // Convert to rpc::TaskEventsData
+  auto data = std::make_unique<rpc::TaskEventData>();
+  data->set_num_profile_task_events_dropped(num_profile_task_events_dropped);
+  data->set_num_status_task_events_dropped(num_status_task_events_dropped);
 
+  size_t num_task_events = to_send.size();
   size_t num_profile_event_to_send = 0;
   size_t num_status_event_to_send = 0;
-  for (auto event : task_events) {
-    if (event.has_profile_events()) {
+  for (auto &task_event : to_send) {
```

Comment:
don't we need to handle the case where the data is empty? I think it is possible with the current implementation

Reply:
E.g., 

buffer size 10, next index 5 batch size 5

1, 2, 3, 4, 5, 6 , 7, 8, 9, 10

report

1, 2, 3, 4, 5

new event 6 (next_index == 6)

1, 2, 3, 4, 5, 6

report

new event 7

1, X, X, X, X, X, 7